src/core/batch.rkt

```diff
@@ -72,163 +76,169 @@
-  (set-batch-roots! out roots)
-  out)
-
-;; Function returns indices of alive nodes within a batch for given roots,
-;;   where alive node is a child of a root + meets a condition - (condition node)
-(define (batch-alive-nodes batch
-                           [roots (batch-roots batch)]
-                           #:keep-vars-alive [keep-vars-alive #f]
-                           #:condition [condition (const #t)])
+  (define brfs
+    (for/list ([expr (in-list exprs)])
+      (batch-add! out expr)))
+  (values out brfs))
+
+(define (batch->progs b brfs)
+  (brfs-belong-to-batch? b brfs)
+  (map (batch-exprs b) brfs))
+
+;; batch-map does not iterate over nodes that are not a child of brf
+;; A lot of parts of Herbie rely on that
+(define (batch-map batch f)
   (define len (batch-length batch))
-  (define alive-mask (make-vector len #f))
-  (for ([root (in-vector roots)])
-    (vector-set! alive-mask root #t))
-  (for ([i (in-range (- len 1) -1 -1)]
-        [node (in-batch batch (- len 1) -1 -1)]
-        [alv (in-vector alive-mask (- len 1) -1 -1)]
-        #:when (or (and alv (condition node)) (and keep-vars-alive (symbol? node))))
-    (unless alv ; if keep-vars-alive then alv may not be #t, making sure it's #t
-      (vector-set! alive-mask i #t))
-    (expr-recurse node
-                  (λ (n)
-                    (when (condition (batch-ref batch n))
-                      (vector-set! alive-mask n #t)))))
-  ; Return indices of alive nodes in ascending order
-  (for/vector ([alv (in-vector alive-mask)]
-               [i (in-naturals)]
-               #:when alv)
-    i))
+  (define out (make-dvector))
+  (define visited (make-dvector))
+  (λ (brf)
+    (match-define (batchref b idx) brf)
+    (cond
+      ; Little check
+      [(or (not (equal? b batch)) (>= idx len)) (error 'batch-map "Inappropriate batchref is passed")]
+      [else
+       (let loop ([idx idx])
+         (match (and (> (dvector-capacity visited) idx) (dvector-ref visited idx))
+           [#t (dvector-ref out idx)]
+           [_
+            (define node (batch-ref batch idx))
+            (define res (f (λ (x) (loop x)) node))
+            (dvector-set! out idx res)
+            (dvector-set! visited idx #t)
+            res]))])))
+
+(define (batch-apply b brfs f)
+  (define out (batch-empty))
+  (define apply-f
+    (batch-map b
+               (λ (remap node)
+                 (define node-with-batchrefs (expr-recurse node (lambda (ref) (batchref b ref))))
+                 (define node* (f node-with-batchrefs))
+                 (define brf*
+                   (let loop ([node* node*])
+                     (match node*
+                       [(? batchref? brf) (remap (batchref-idx brf))]
+                       [_ (batch-push! out (expr-recurse node* (compose batchref-idx loop)))])))
+                 brf*)))
+  (define brfs* (map apply-f brfs))
+  (values out brfs*))
+
+(define (batch-apply! b f)
```

Comment:
So I think the difference between `batch-apply!` and `batch-map` are:

- `batch-apply!` wraps references with batchrefs, `batch-map` doesn't?
- `batch-apply!` automatically rewrites subnodes, `batch-map` doesn't?
- `batch-apply!` requires outputs to be nodes, `batch-map` is generic
- `batch-apply!` adds to the same batch

Not against a helper function but some of these seem like generally good features? My proposal is that `batch-map` also wraps references with batchrefs, and `batch-apply!` not rewrite subnodes (pass `remap` to `f` and require the user to call it). Then the two functions are more similar and it's clearer why you'd use one or the other.